tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/CustomGradient.java

```diff
@@ -0,0 +1,55 @@
+/*
+ Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================
+*/
+package org.tensorflow.op;
+
+import java.util.List;
+import org.tensorflow.Operand;
+import org.tensorflow.Output;
+import org.tensorflow.TensorFlow;
+import org.tensorflow.internal.c_api.GradFunc;
+
+/**
+ * A custom gradient for ops of type {@link T}. Should be registered using {@link
+ * TensorFlow#registerCustomGradient(Class, CustomGradient)}.
+ *
+ * @param <T> the type of op this gradient is for.
+ */
+@SuppressWarnings("rawtypes")
+@FunctionalInterface
+public interface CustomGradient<T extends RawOpInputs> {
```

Comment:
Could this be `CustomGradient<T extends RawOpInputs<?>>` to get rid of the rawtypes warning?

Reply:
I tried that at first, but it causes "cannot infer T" errors when `registerCustomGradient` is used and I have no idea why.

Reply:
What's the exact error you get?

Reply:
```
/C:/Users/jimne/Desktop/OtherStuff/tensorflow_java/tensorflow-core/tensorflow-core-api/src/test/java/org/tensorflow/CustomGradientTest.java:[57,44] incompatible types: cannot infer type-variable(s) T
    (argument mismatch; cannot infer functional interface descriptor for org.tensorflow.op.CustomGradient<org.tensorflow.op.nn.NthElement.Inputs>)
```

Reply:
Interestingly, the CI reports it as:
```
Error:  /__w/java/java/tensorflow-core/tensorflow-core-api/src/test/java/org/tensorflow/CustomGradientTest.java:[57,44] incompatible types: null
```

Reply:
So I think that's because the raw type is washing off all the generics, so javac compiles it. Which means that this functional interface isn't actually doing any generic type checking.

Reply:
Yeah, makes sense.  Do you know of a way around the usage error?  As far as I can tell it should work ok even without the rawtypes, but it doesn't.